File: new-resource-detector/src/test/java/nl/haploid/octowight/service/EventConsumerServiceIT.java
package nl.haploid.octowight.service;

import nl.haploid.octowight.AbstractIT;
import nl.haploid.octowight.AtomChangeEvent;
import nl.haploid.octowight.JsonMapper;
import nl.haploid.octowight.TestData;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.Timeout;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

import static org.junit.Assert.assertEquals;

public class EventConsumerServiceIT extends AbstractIT {

	@Autowired
	private EventConsumerService service;

	@Autowired
	private KafkaProducer<String, String> kafkaProducer;

	@Autowired
	private JsonMapper jsonMapper;

	@Rule
	public Timeout globalTimeout = new Timeout(10, TimeUnit.SECONDS);

	@Test
	public void testConsumeMessage() throws InterruptedException, ExecutionException {
		final String topic = TestData.topic();
		service.setTopic(topic);
		final AtomChangeEvent expectedEvent = TestData.atomChangeEvent("joan");
		sendMessage(topic, expectedEvent);
		final AtomChangeEvent actualEvent = service.consumeMessage();
		assertEquals(expectedEvent, actualEvent);
	}

	public void sendMessage(final String topic, final AtomChangeEvent event) throws InterruptedException, ExecutionException {
		final String message = jsonMapper.toString(event);
		log.debug(String.format("Send message: %s", message));
		final ProducerRecord<String, String> record = new ProducerRecord<>(topic, message);
		kafkaProducer.send(record).get();
	}

	@Test
	public void testConsumeMessages() throws InterruptedException, ExecutionException {
		final String topic = TestData.topic();
		service.setTopic(topic);
		final AtomChangeEvent event1 = TestData.atomChangeEvent("bob");
		final AtomChangeEvent event2 = TestData.atomChangeEvent("benson");
		final List<AtomChangeEvent> expectedEvents = Arrays.asList(event1, event2);
		sendMessage(topic, event1);
		sendMessage(topic, event2);
		final List<AtomChangeEvent> actualEvents = service.consumeMessages(10)
				.collect(Collectors.toList());
		assertEquals(expectedEvents, actualEvents);
	}

	@Test
	public void testCommit() throws ExecutionException, InterruptedException {
		final String topic = TestData.topic();
		service.setTopic(topic);
		final AtomChangeEvent event = TestData.atomChangeEvent("harris");
		sendMessage(topic, event);
		service.consumeMessages(10)
				.collect(Collectors.toList());
		service.commit();
		service.reset();
		final List<AtomChangeEvent> actualEvents = service.consumeMessages(10)
				.collect(Collectors.toList());
		assertEquals(0, actualEvents.size());
	}
}

File: event-consumer/src/test/scala/nl/haploid/octowight/service/EventConsumerServiceIT.scala
package nl.haploid.octowight.service

import java.util.concurrent.TimeUnit

import nl.haploid.octowight.{AbstractIT, AtomChangeEvent, JsonMapper, TestData}
import org.apache.kafka.clients.producer.{KafkaProducer, ProducerRecord}
import org.junit.Rule
import org.junit.rules.Timeout
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired

class EventConsumerServiceIT extends AbstractIT {
  private[this] lazy val log = LoggerFactory.getLogger(getClass)

  @Autowired private[this] val eventConsumerService: EventConsumerService = null
  @Autowired private[this] val kafkaProducer: KafkaProducer[String, String] = null
  @Autowired private[this] val jsonMapper: JsonMapper = null

  @Rule val globalTimeout = new Timeout(10, TimeUnit.SECONDS)

  def sendMessage(topic: String, event: AtomChangeEvent): Unit = {
    val message = jsonMapper.serialize(event)
    log.debug(s"Send message: $message")
    val record = new ProducerRecord[String, String](topic, message)
    kafkaProducer.send(record).get
  }

  "Event consumer" should "consume a message" in {
    val topic = TestData.topic
    eventConsumerService.reset(topic)
    val expectedEvent = TestData.atomChangeEvent("joan")
    sendMessage(topic, expectedEvent)
    val actualEvent = eventConsumerService.consumeEvent
    actualEvent should be(expectedEvent)
  }

  "Event consumer" should "consume multiple messages" in {
    val topic = TestData.topic
    eventConsumerService.reset(topic)
    val event1 = TestData.atomChangeEvent("bob")
    val event2 = TestData.atomChangeEvent("benson")
    val expectedEvents = Set(event1, event2)
    sendMessage(topic, event1)
    sendMessage(topic, event2)
    val actualEvents = eventConsumerService.consumeDistinctEvents()
    actualEvents should be(expectedEvents)
  }

  "Event consumer" should "commit stream offset" in {
    val topic = TestData.topic
    eventConsumerService.reset(topic)
    val event = TestData.atomChangeEvent("harris")
    sendMessage(topic, event)
    eventConsumerService.consumeDistinctEvents()
    eventConsumerService.commit()
    eventConsumerService.reset(topic)
    val actualEvents = eventConsumerService.consumeDistinctEvents()
    actualEvents should have size 0
  }
}
